src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,635 @@
+
+soter_status_t soter_engine_specific_to_rsa_pub_key(const soter_engine_specific_rsa_key_t* engine_key,
+                                                    soter_container_hdr_t* key,
+                                                    size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    int rsa_mod_size;
+    size_t output_length;
+    uint32_t* pub_exp;
+    // This bigint is reused multiple times: read into it, serialize it, read into it...
+    BIGNUM* tmp = NULL;
+
+    if (!key_length) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (EVP_PKEY_RSA != EVP_PKEY_id(pkey)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    // Relevant pkey params
+    // OSSL_PKEY_PARAM_BITS, "bits", int
+    // OSSL_PKEY_PARAM_RSA_N, "n", big uint
+    // OSSL_PKEY_PARAM_RSA_E, "e", big uint
+
+    // See https://docs.cossacklabs.com/themis/spec/asymmetric-keypairs/rsa/ for key layout
+
+    if (!EVP_PKEY_get_int_param(pkey, OSSL_PKEY_PARAM_BITS, &rsa_mod_size)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+    rsa_mod_size = (rsa_mod_size + 7) / 8;
+
+    if (!is_mod_size_supported(rsa_mod_size)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    output_length = rsa_pub_key_size(rsa_mod_size);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_N, &tmp)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (BN_bn2binpad(tmp, (unsigned char*)(key + 1), rsa_mod_size) == -1) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_E, &tmp)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    pub_exp = (uint32_t*)((unsigned char*)(key + 1) + rsa_mod_size);
+    if (BN_is_word(tmp, RSA_F4)) {
+        *pub_exp = htobe32(RSA_F4);
+    } else if (BN_is_word(tmp, RSA_3)) {
+        *pub_exp = htobe32(RSA_3);
+    } else {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    memcpy(key->tag, rsa_pub_key_tag(rsa_mod_size), SOTER_CONTAINER_TAG_LENGTH);
+    key->size = htobe32(output_length);
+    soter_update_container_checksum(key);
+    *key_length = output_length;
+    res = SOTER_SUCCESS;
+
+err:
+    BN_free(tmp);
+
+    return res;
+}
+
+soter_status_t soter_engine_specific_to_rsa_priv_key(const soter_engine_specific_rsa_key_t* engine_key,
+                                                     soter_container_hdr_t* key,
+                                                     size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    int rsa_mod_size;
+    size_t output_length;
+    uint32_t* pub_exp;
+    // This bigint is reused multiple times: read into it, serialize it, read into it...
+    BIGNUM* tmp = NULL;
+    unsigned char* curr_bn = (unsigned char*)(key + 1);
```

Comment:
same here. can you review all functions and add checks for input arguments?

Reply:
This one is kinda weird. Indeed, we assign `curr_bn = key` before checking key. But we access `curr_bn` only after checking `key`, so there is no bug here. To make it cleaner, I'll move assignment to a place between check and first usage